app/src/org/commcare/android/view/c3/DataConfiguration.java

```diff
@@ -0,0 +1,349 @@
+package org.commcare.android.view.c3;
+
+import org.commcare.android.util.InvalidStateException;
+import org.commcare.suite.model.graph.AnnotationData;
+import org.commcare.suite.model.graph.BubblePointData;
+import org.commcare.suite.model.graph.Graph;
+import org.commcare.suite.model.graph.GraphData;
+import org.commcare.suite.model.graph.SeriesData;
+import org.commcare.suite.model.graph.XYPointData;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.Vector;
+
+/**
+ * Data-related configuration for C3. This configuration should be run before
+ * any others, as th data will sometimes affect other configuration.
+ *
+ * Created by jschweers on 11/16/2015.
+ */
+public class DataConfiguration extends Configuration {
+    // Actual data: array of arrays, where first element is a string id
+    // and later elements are data, either x values or y values.
+    private JSONArray mColumns = new JSONArray();
+
+    // Hash that pairs up the arrays defined in columns,
+    // y-values-array-id => x-values-array-id
+    private JSONObject mXs = new JSONObject();
+
+    // Hash of y-values id => name for legend
+    JSONObject mNames = new JSONObject();
+
+    // Hash of y-values id => 'y' or 'y2' depending on whether this data
+    // should be plotted against the primary or secondary y axis
+    private JSONObject mAxes = new JSONObject();
+
+    // Hash of y-values id => line, scatter, bar, area, etc.
+    JSONObject mTypes = new JSONObject();
+
+    // Hash of y-values id => series color
+    JSONObject mColors = new JSONObject();
+
+    public DataConfiguration(GraphData data) throws JSONException, InvalidStateException {

```

Comment:
I'd rather not split into subclasses because the future direction of graphing is to mix different types of series in the same graph (admittedly, there's other code in this PR that expects type to be graph-level, but I'd prefer not to make even more). Agree that this is messy, though, will clean it up.